File: algorithm/LinkedList/0002_add_two_numbers.ts
/**
 * @author lizhi.guo@foxmail.com
 * @source https://leetcode-cn.com/problems/add-two-numbers/
 * @time   2019-10-25
 */

type ListNodeType = ListNode | null;

class ListNode {
  val: number;
  next: ListNodeType;

  constructor(val: number) {
    this.val = val;
    this.next = null;
  }
}

// 创建新的结果链表，时间复杂度为O(n)，空间复杂度为O(n)
const addTwoNumbers = function(l1: ListNodeType, l2: ListNodeType): ListNodeType {
  let dummyHead = new ListNode(0);
  let cur = dummyHead;
  let carry = 0; // 用来存储超过10的部分

  while(l1 !== null || l2 !== null) {
    let x = (l1 !== null) ? l1.val : 0;
    let y = (l2 !== null) ? l2.val : 0;
    let sum = x + y + carry;

    carry = Math.floor(sum / 10);
    cur.next = new ListNode(sum % 10);
    cur = cur.next;

    if (l1 !== null) {
      l1 = l1.next;
    }

    if (l2 !== null) {
      l2 = l2.next;
    }
  }

  if (carry > 0) {
    cur.next = new ListNode(carry);
  }

  return dummyHead.next;
}

// 使用l1作为结果链表，时间复杂度为O(n)，空间复杂度为O(n)
const addTwoNumbers2 = function(l1: ListNodeType, l2: ListNodeType): ListNodeType {
  let p1 = l1;
  let p2 = l2;
  let prev = new ListNode(0);
  let carry = 0;

  while (p1 !== null || p2 !== null) {
    let a = p1 ? p1.val : 0;
    let b = p2 ? p2.val : 0;
    if (p1) {
      p1.val = (a + b + carry) % 10;
    } else {
      prev.next = new ListNode((a + b + carry) % 10);
      p1 = prev.next;
    }

    carry = Math.floor((a + b + carry) / 10);
    prev = p1;
    p1 = p1.next;
    if (p2) {
      p2 = p2.next;
    }
  }

  prev.next = carry ? new ListNode(1) : null;
  return l1;
}

File: Leetcode/Go/0002-add-two-numbers.go
package leetcode

// ListNode 链表节点
type ListNode struct {
	Val  int
	Next *ListNode
}

/*
	迭代的方法：每次都需要判断当前l1,l2是否为nil，并且余数carry是否已经置零
	时间复杂度O(n)
	空间复杂度O(1)
*/
func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	// 定义一个虚拟头结点
	dummyHead := &ListNode{}
	currNode := dummyHead
	// 用来存储当前两数相加的余数
	carry := 0
	for l1 != nil || l2 != nil || carry != 0 {
		currSum := carry
		if l1 != nil {
			currSum += l1.Val
			l1 = l1.Next
		}
		if l2 != nil {
			currSum += l2.Val
			l2 = l2.Next
		}
		currNode.Next = &ListNode{Val: currSum % 10}
		currNode = currNode.Next
		if currSum >= 10 {
			carry = 1
		} else {
			carry = 0
		}
	}
	return dummyHead.Next
}

/*
	递归方法
	时间复杂度：O(n)
	空间复杂度：O(n)
*/
func addTwoNumbers2(l1 *ListNode, l2 *ListNode) *ListNode {
	return dfs(l1, l2, 0)
}

// carry 每次进行两数相加的除以10之后的余数
func dfs(l1 *ListNode, l2 *ListNode, carry int) *ListNode {
	// 递归终止条件
	if l1 == nil && l2 == nil && carry == 0 {
		return nil
	}
	currSum := carry
	var l1Next, l2Next *ListNode
	if l1 != nil {
		currSum += l1.Val
		l1Next = l1.Next
	}
	if l2 != nil {
		currSum += l2.Val
		l2Next = l2.Next
	}
	var result = &ListNode{Val: currSum % 10}
	if currSum >= 10 {
		carry = 1
	} else {
		carry = 0
	}
	result.Next = dfs(l1Next, l2Next, carry)
	return result
}
